File: frontend/src/api/referral.ts
/**
 * Referral API Client
 * API клиент для работы с реферальной программой
 */

import apiClient from './client';

/**
 * Информация о реферале
 */
export interface ReferralItem {
  id: number;
  telegram_id: number;
  username: string | null;
  credits_awarded: number;
  is_awarded: boolean;
  created_at: string;
}

/**
 * Ответ с реферальной ссылкой
 */
export interface ReferralLinkResponse {
  referral_link: string;
  referral_code: string;
  total_referrals: number;
  total_earned: number;
}

/**
 * Запрос на регистрацию по реферальному коду
 */
export interface ReferralRegisterRequest {
  referral_code: string;
}

/**
 * Ответ после регистрации по реферальному коду
 */
export interface ReferralRegisterResponse {
  success: boolean;
  message: string;
  bonus_credits: number;
}

/**
 * Статистика по рефералам
 */
export interface ReferralStatsResponse {
  total_referrals: number;
  active_referrals: number;
  pending_referrals: number;
  total_earned: number;
  referrals: ReferralItem[];
  referral_link: string;
  referral_code: string;
}

/**
 * Получить реферальную ссылку текущего пользователя
 * @returns Реферальная ссылка и статистика
 */
export const getReferralLink = async (): Promise<ReferralLinkResponse> => {
  const response = await apiClient.get<ReferralLinkResponse>(
    '/api/v1/referrals/link'
  );
  return response.data;
};

/**
 * Зарегистрировать пользователя по реферальному коду
 * @param request - Реферальный код
 * @returns Результат регистрации
 */
export const registerReferral = async (
  request: ReferralRegisterRequest
): Promise<ReferralRegisterResponse> => {
  const response = await apiClient.post<ReferralRegisterResponse>(
    '/api/v1/referrals/register',
    request
  );
  return response.data;
};

/**
 * Получить статистику по рефералам
 * @returns Статистика: количество рефералов, заработанные кредиты, список рефералов
 */
export const getReferralStats = async (): Promise<ReferralStatsResponse> => {
  const response = await apiClient.get<ReferralStatsResponse>(
    '/api/v1/referrals/stats'
  );
  return response.data;
};

/**
 * Скопировать реферальную ссылку в буфер обмена
 * @param link - Реферальная ссылка
 * @returns true если успешно скопировано
 */
export const copyReferralLink = async (link: string): Promise<boolean> => {
  try {
    // Используем Telegram WebApp API для копирования, если доступно
    if (window.Telegram?.WebApp) {
      window.Telegram.WebApp.showAlert(`Ссылка скопирована:\n${link}`);
    }

    // Копируем в буфер обмена
    await navigator.clipboard.writeText(link);
    return true;
  } catch (error) {
    console.error('Failed to copy referral link:', error);
    return false;
  }
};

/**
 * Поделиться реферальной ссылкой через Telegram
 * @param link - Реферальная ссылка
 * @param text - Текст сообщения
 */
export const shareReferralLink = (link: string, text?: string): void => {
  const message = text || `Присоединяйся к AI Image Generator!\n${link}`;

  // Используем Telegram Share API
  if (window.Telegram?.WebApp) {
    const url = `https://t.me/share/url?url=${encodeURIComponent(
      link
    )}&text=${encodeURIComponent(message)}`;
    window.Telegram.WebApp.openTelegramLink(url);
  } else {
    // Fallback: открываем в новом окне
    const url = `https://t.me/share/url?url=${encodeURIComponent(
      link
    )}&text=${encodeURIComponent(message)}`;
    window.open(url, '_blank');
  }
};

File: backend/app/schemas/referral.py
"""
Pydantic схемы для реферальной программы
"""

from datetime import datetime
from pydantic import BaseModel, Field


class ReferralLinkResponse(BaseModel):
    """Ответ с реферальной ссылкой"""

    referral_link: str = Field(..., description="Реферальная ссылка для приглашения друзей")
    referral_code: str = Field(..., description="Реферальный код пользователя")
    total_referrals: int = Field(..., description="Общее количество приглашённых друзей")
    total_earned: int = Field(..., description="Всего заработано ⭐️звезд за рефералов")


class ReferralRegisterRequest(BaseModel):
    """Запрос на регистрацию по реферальной ссылке"""

    referral_code: str = Field(
        ...,
        description="Реферальный код пригласившего пользователя",
        min_length=1,
        max_length=100,
    )


class ReferralRegisterResponse(BaseModel):
    """Ответ после регистрации по реферальной ссылке"""

    success: bool = Field(..., description="Успешно ли зарегистрирован реферал")
    message: str = Field(..., description="Сообщение о результате")
    bonus_credits: int = Field(
        default=0, description="Количество бонусных ⭐️звезд (если есть)"
    )


class ReferralItem(BaseModel):
    """Информация об одном реферале"""

    id: int = Field(..., description="ID реферала")
    telegram_id: int | None = Field(
        None, description="Telegram ID приглашённого пользователя"
    )
    username: str | None = Field(None, description="Username приглашённого пользователя")
    credits_awarded: int = Field(..., description="Количество начисленных ⭐️звезд")
    is_awarded: bool = Field(
        ..., description="Были ли начислены ⭐️звезды (после первого действия)"
    )
    created_at: datetime = Field(..., description="Дата приглашения")

    class Config:
        from_attributes = True


class ReferralStatsResponse(BaseModel):
    """Статистика по рефералам"""

    total_referrals: int = Field(..., description="Общее количество приглашённых друзей")
    active_referrals: int = Field(
        ..., description="Активные рефералы (совершили хотя бы одно действие)"
    )
    pending_referrals: int = Field(
        ..., description="Ожидающие рефералы (ещё не совершили действие)"
    )
    total_earned: int = Field(..., description="Всего заработано ⭐️звезд")
    referrals: list[ReferralItem] = Field(
        default_factory=list, description="Список рефералов"
    )
    referral_link: str = Field(..., description="Реферальная ссылка")
    referral_code: str = Field(..., description="Реферальный код")


class ReferralReward(BaseModel):
    """Информация о награде за реферала"""

    referral_id: int = Field(..., description="ID реферальной записи")
    credits_awarded: int = Field(..., description="Количество начисленных ⭐️звезд")
    referrer_id: int = Field(..., description="ID пользователя, получившего награду")
    referred_id: int = Field(..., description="ID пользователя, который был приглашён")
